test_conformance/extensions/cl_khr_command_buffer/cl_khr_command_buffer_mutable_dispatch/mutable_command_work_groups.cpp

```diff
@@ -0,0 +1,275 @@
+//
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <extensionHelpers.h>
+#include "imageHelpers.h"
+#include "mutable_command_basic.h"
+
+#include <CL/cl.h>
+#include <CL/cl_ext.h>
+
+////////////////////////////////////////////////////////////////////////////////
+// mutable dispatch tests which handle following cases:
+//
+// 1. The command buffer is created with CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR in its properties.
+// 2. The ND-range command is recorded with CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR in its properties.
+// 3. Both the command buffer and ND-range command have CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR in their properties.
+
+typedef struct {
+    const cl_command_buffer_properties_khr *command_buffer_properties;
+    const cl_ndrange_kernel_command_properties_khr *ndrange_properties;
+} Configuration;
+
+// Define the command buffer properties for each configuration
+const cl_command_buffer_properties_khr command_buffer_properties[] = {
+        CL_COMMAND_BUFFER_MUTABLE_DISPATCH_ASSERTS_KHR,
+        CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR,
+        0
+};
+
+// Define the ndrange properties
+const cl_ndrange_kernel_command_properties_khr ndrange_properties[] = {
+        CL_MUTABLE_DISPATCH_UPDATABLE_FIELDS_KHR,
+        CL_MUTABLE_DISPATCH_GLOBAL_SIZE_KHR,
+        CL_MUTABLE_DISPATCH_ASSERTS_KHR,
+        CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR,
+        0
+};
+// Initialize the array of configurations
+Configuration configurations[] = {
+        {command_buffer_properties, nullptr},
+        {nullptr,                   ndrange_properties},
+        {command_buffer_properties, ndrange_properties}
+};
+
+template<int test_case>
+struct MutableDispatchWorkGroups : public BasicMutableCommandBufferTest {
+
+    MutableDispatchWorkGroups(cl_device_id device, cl_context context,
+                              cl_command_queue queue)
+            : BasicMutableCommandBufferTest(device, context, queue), out_of_order_queue(nullptr), command_buffer(this) {
+        out_of_order_support = true;
+        config = configurations[test_case];
+    }
+
+    bool Skip() override {
```

Comment:
We should be calling base class Skip in here, `BasicMutableCommandBufferTest::Skip()`

Reply:
The issue here is that our implementation requires the following properties for command buffer queues - `CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE | CL_QUEUE_NO_SYNC_OPERATIONS_INTEL`. When we developed this test, we wanted it to be able to run on our runtime; this is why we have this logic in `Skip` and `Setup` (all other command buffer tests are skipped with our implementation).

Reply:
I understand, but I don't really like forcing the out-of-order requirement on a cross vendor CTS test. I wonder if we could parameterize the test implementation so that there's both an in-order queue test and out-of-order queue test? Since there is only one command the difference should be minimal, and it would mean `CL_MUTABLE_DISPATCH_ASSERT_NO_ADDITIONAL_WORK_GROUPS_KHR` would get test coverage both for yourself and any implementers that don't support out-of-order queues.

As an aside, it would be much better if we could find a way for you to run all the command-buffer tests, but that feels out of scope for this PR.

Reply:
We just discussed this issue internally and updated the test to match the structure of the other sub-tests

Reply:
My last comment is tangential to the original one about test structure, so i've created a new thread for it https://github.com/KhronosGroup/OpenCL-CTS/pull/1991/files#r1668852431